Readme.md
[![Build Status](https://secure.travis-ci.org/bodenr/errors.png)](http://travis-ci.org/bodenr/errors)
# errors
    
Errors is a comprehensive and robust, yet lightweight, set of error utilities for 
[node.js](http://nodejs.org) enabling you to _do_ errors more effectively.

## Features

* Parameterized error factory allowing you do define how errors should behave 
based on your project needs.
* Support for enterprise level error attributes including more detailed error cause
and operator response messages.
* Predefined error constructors for all HTTP 4xx-5xx based errors allowing you to leverage
HTTP errors out of the box.
* Error mapping via registered mapping function permitting you to map between 
errors when needed.

## Installation

Install using `npm`:

    $ npm install @bubblehunt/errors

## Running the tests

From the `errors` directory first install the dev dependencies:
```
npm install
```

Then run the tests:
```
npm test
```

## API Docs

The API docs are provided in html and md format and are located under
`errors/docs/`. If you want to rebuild them for any reason, you can
run the following from the `errors` directory:
```
make doc
```

## Defining error messages

The examples assume you've `require`d the errors module like so:
```js
const errors = require('@bubblehunt/errors').default;
```

or
```js
import errors from '@bubblehunt/errors';
```

#### Override flow types

See example at `test/flow`


#### Create new errors

Create a very barebones error -- you must specify at least the error name:

```js
// barebones
errors.create({name: 'RuntimeError'});
console.log(new errors.RuntimeError().toString());
```

produces:
```
RuntimeError: An unexpected RuntimeError occurred.
Code: 601
```

You can define a default message for the error:

```js
// default message
errors.create({
    name: 'RuntimeError',
    message: 'A runtime error occurred during processing'
});
console.log(new errors.RuntimeError().toString());
```

which outputs:
```
RuntimeError: A runtime error occurred during processing
Code: 602

```

Define a default message, explanation and response:

```js
// default message, explanation and response
errors.create({
    name: 'FileNotFoundError',
    message: 'The requested file could not be found',
    explanation: 'The file /home/boden/foo could not be found',
    response: 'Verify the file exists and retry the operation'
});
console.log(new errors.FileNotFoundError().toString());
```

gives us:
```
FileNotFoundError: The requested file could not be found
Code: 603
Explanation: The file /home/boden/foo could not be found
Response: Verify the file exists and retry the operation

```

Override messages on instantiation:
```js
// override messages
console.log(new errors.FileNotFoundError(
        'Cannot read file'
        , 'You do not have access to read /root/foo'
        , 'Request a file you have permissions to access').toString());
```

outputs:
```
FileNotFoundError: Cannot read file
Code: 603
Explanation: You do not have access to read /root/foo
Response: Request a file you have permissions to access

```

Use the options style constructor to assign standard properties:
```js
console.log(new errors.Http401Error({
	message: "Expired Token",
	explanation: "Your token has expired"}).toString());
```

outputs:
```
Http401Error: Expired Token
Code: 401
Explanation: Your token has expired
Error: Expired Token
```

Using the options style constructor you can also assign
arbitrary non-standard properties:
```js
console.log(new errors.Http401Error({
	message: "Expired Token",
	explanation: "Your token has expired",
	expired: new Date()}).toString());
```

outputs:
```
Http401Error: Expired Token
Code: 401
Explanation: Your token has expired
expired: Fri Jun 20 2014 04:19:41 GMT-0400 (EDT)
```

Note however that you cannot assign values to the
`stack`, `name` or `code` standard property:
```js
console.log(new errors.Http401Error({
	name: "ExpiredToken"}).toString());
```

outputs:
```
/home/boden/workspace/errors/lib/errors.js:261
    			throw Error("Properties 'stack', 'name' or 'code' " +
    			      ^
Error: Properties 'stack', 'name' or 'code' cannot be overridden
    at Error (<anonymous>)
    at new scope.(anonymous function) (/home/boden/workspace/errors/lib/errors.js:261:14)
    at Object.<anonymous> (/home/boden/workspace/errors/examples/basic/usage.js:126:13)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:906:3
```

## Error codes

If you don't provide a `code` when defining the error, a unique code will
be assigned for you. Unique codes start at 600 and increase by 1 for each
error defined.

If you prefer to manage your own error codes, for example to group related
errors into blocks of codes, just specify a `code`:
```js
// define code
errors.create({
    name: 'SecurityError',
    code: 1100
});
console.log(new errors.SecurityError().toString());
```

which logs:
```
SecurityError: An unexpected SecurityError occurred.
Code: 1100

```

## Inheritance

You can build a hierarchy of errors by specifying the `parent` your
error should inherit from. If no `parent` is specified, the error
will inherit from `Error`.

For example:
```js
// inheritance
errors.create({
    name: 'FatalError',
    message: 'A fatal error occurred',
});
errors.create({
    name: 'FatalSecurityError',
    message: 'A security error occurred, the app must exit',
    parent: errors.FatalError
});
try {
    throw new errors.FatalSecurityError();
} catch (e) {
    if (e instanceof errors.FatalError) {
        // exit
        console.log("Application is shutting down...");
    }
}
```

will produce:
```
Application is shutting down...
```

## Namespacing

By default, newly defined errors are created on the `exports` of
the errors module, but you can specify where the error should
be defined.

For example to define an error on your module's `exports`:
```js
// namespace
errors.create({
    name: 'MalformedExpressionError',
    scope: exports
});
console.log(new exports.MalformedExpressionError().toString());
```

## Looking up errors

For convenience, errors keeps track of all the errors you've defined
via the errors module and allows you to look them up via `name` or
`code`.

So from our previous example:
```js
errors.find(1100);
errors.find('SecurityError')
```

Will both return the `SecutiryError` we defined.

## Stack traces

By default stack traces are disabled which means that error methods
like `toString()` and `toJSON()` return representation without stack 
traces. You can enable stack traces by leveraging the `errors.stacks()`
method.

For example:
```js
errors.stacks(true);
new errors.Http413Error().toString();
// => includes stack trace
new errors.Http413Error().toJSON();
// => includes a 'stack' property
```

You can also use the `errors.stacks()` method without arguments to 
retrieve the current value of stacks.

This allows you to write code like:
```js
if (errors.stacks()) {
    // => stack traces enabled
}
```

## Mappers

You can register and leverage mapper functions which allow you to
map from one (or more) error types into another.

For example if you wanted to mask invalid user and password errors into
a generic credentials error:
```js
// mappers
errors.create({name: 'InvalidUsernameError'});
errors.create({name: 'InvalidPasswordError'});
errors.mapper(['InvalidUsernameError', 'InvalidPasswordError'], function(err) {
    return new errors.SecurityError('Invalid credentials supplied');
});
console.log(errors.mapError(new errors.InvalidUsernameError()).toString());
console.log(errors.mapError(new errors.InvalidPasswordError()).toString());
```

outputs:
```
SecurityError: Invalid credentials supplied
Code: 1100
SecurityError: Invalid credentials supplied
Code: 1100

```

## Native errors

Often times you need to extract 'errors-like' properties from native error
objects. For example you have a native JS or node error and you want to 
extract it's errors-like properties. An error's module-level function
called `errors.errorToJSON()` allows you to do this.

For example to extract error properties from a native error (`errors.stacks`
is set to `false` in this example):
```js
console.log("%j", errors.errorToJSON(new TypeError("Bad type")));
```

outputs:
```
{"message":"Bad type","name":"TypeError"}
```

You can also remap error attributes which may be nested. For example:

```js
console.log("%j", errors.errorToJSON(new TypeError("Bad type"), 
    {'className': ['constructor.name'], 'message': ['message']}));
```

outputs:
```
{"className":"TypeError","message":"Bad type"}
```


## Predefined HTTP 4xx-5xx errors

The errors module predefines a set of errors which represent HTTP
4xx-5xx responses. These errors are exported by the errors module and use the
naming convention `Http[code]Error`. For example `Http401Error` and 
`Http500Error` which have a code of `401` and `500` respectively.

For example to leverage the HTTP errors:
```js
throw new errors.Http401Error();
// ...
throw new errors.Http500Error('Something bad happened');
```


## License

(The MIT License)

Copyright (c) 2012 Boden Russell &lt;bodensemail@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
0.8.0 / 2018-04-20
==================

* Replace defaultMessage, defaultExplanation and defaultResponse with message, explanation and response
* Revisioned flow types


0.5.0 / 2018-04-17
==================

* Add flow and es6
* Drop dependencies
* Remove express.js and connect.js to make it more universal


0.0.1 / 2012-11-06
==================

* Initial publication


test/flow.js
/* @flow */

// =============================================================================
// Imports
// =============================================================================
import
    Errors, {
        type ErrorsModuleType,
        type DefaultErrorType,
        type DefaultErrorOptionsType,
    }
from '../src/errors';

// Flow types
// --------------------------------------------------------

type MyErrorPropsType = {
    _message?: string,
    myProp?: number,
};

/* To allow any prop
type MyErrorPropsType = Object;
*/

type MyErrorType =
    & DefaultErrorType
    & MyErrorPropsType;

type MyOptionsType =
    & DefaultErrorOptionsType
    & MyErrorPropsType;


// =============================================================================
// Constants
// =============================================================================
const MyErrors: ErrorsModuleType<MyErrorType, MyOptionsType> = Errors;


// =============================================================================
// Tests
// =============================================================================

// Correct
// --------------------------------------------------------
var MyError = MyErrors.create({
    name: 'MYERROR',
    _message: 'some message',
    myProp: 1,
});

var myError = new MyError();

myError._message
myError.myProp

myError.code
myError.status
myError.response
myError.explanation

// Errors
// --------------------------------------------------------
// $FlowThrowsError
myError._message1

MyErrors.create({
    name: 'MYERROR',
    // $FlowThrowsError
    _message: 1,
});

MyErrors.create({
    name: 'MYERROR',
    // $FlowThrowsError
    myProp: '1',
});




test/test.js

var errors = require('../dist/errors').default
    , http = require('http')
    , should = require('should');

describe('errors export structure', function() {
    it('should contain find() method', function() {
        errors.should.be.a('object').and.have.property('find');
        errors.find.should.be.a('function');
    });

    it('should contain a create method', function() {
        errors.create.should.be.a('function');
        errors.create.should.have.length(1);
    });

    it('should contain HttpError', function() {
        errors.HttpError.should.be.a('function');
        var httpError = new errors.HttpError();
        httpError.should.an.instanceOf(Error);
        httpError.code.should.be.above(599);
    });

    it('should not leak to global.code', function() {
        should.not.exist(global.code);
    });

    for (var code in http.STATUS_CODES) {
        it('should contain ' + http.STATUS_CODES[code] + ' Error', function() {
            code = parseInt(code);
            var errorName = 'Http' + code + 'Error'
                , errorInstance = new errors[errorName]();
            errors[errorName].should.be.a('function');
            errorInstance.should.be.an.instanceOf(errors.HttpError);
            errorInstance.code.should.equal(code);
            errorInstance.message.should.equal(http.STATUS_CODES[code]);
            errorInstance.status.should.equal(code);
        });
    }
});

var FatalError = errors.create({
    name: 'FatalError',
    message: 'A Fatal Error Occurred.'
});
var FatalDBError = errors.create({
    name: 'FatalDBError',
    parent: FatalError,
    message: 'A Fatal Database Error Occurred.'
});
var FatalDBTransactionError = errors.create({
    name: 'FatalDBTransactionError',
    parent: FatalDBError,
    message: 'A Fatal Database Transaction Error Occurred.'
});

var fatalError = new FatalError()
    , fatalDBError = new FatalDBError()
    , fatalDBTransError = new FatalDBTransactionError();

describe('errors inheritance', function() {

    it('FatalError extends Error', function() {
        errors.FatalError.should.equal(FatalError);
        FatalError.should.be.a('function');
        fatalError.should.be.an.instanceOf(Error);
        fatalError.code.should.be.above(599);
        fatalError.message.should.equal('A Fatal Error Occurred.');
        fatalError.status.should.equal(500);
    });

    it('FatalDBError extends FatalError', function() {
        errors.FatalDBError.should.equal(FatalDBError);
        FatalDBError.should.be.a('function');
        fatalDBError.should.be.an.instanceOf(FatalError);
        fatalDBError.should.not.be.an.instanceOf(FatalDBTransactionError);
        fatalDBError.code.should.be.above(599);
        fatalDBError.code.should.not.equal(new FatalError().code);
        fatalDBError.message.should.equal('A Fatal Database Error Occurred.');
        fatalDBError.status.should.equal(500);
    });

    it('FatalDBTransactionError extends FatalDBError', function() {
        errors.FatalDBTransactionError.should.equal(FatalDBTransactionError);
        FatalDBTransactionError.should.be.a('function');
        fatalDBTransError.should.be.an.instanceOf(FatalDBError);
        fatalDBTransError.code.should.be.above(599);
        fatalDBTransError.code.should.not.equal(new FatalDBError().code);
        fatalDBTransError.message.should.equal('A Fatal Database Transaction Error Occurred.');
        fatalDBTransError.status.should.equal(500);
    });
});

describe('errors.find()', function() {

    it('should find existing error by code', function() {
        errors.find(fatalDBError.code).should.equal(FatalDBError);
    });

    it('should find existing error by name', function() {
        errors.find(fatalDBError.name).should.equal(FatalDBError);
    });

    it('should not find error for non-existing code', function() {
        require('should').not.exist(errors.find(9999));
    });

    it('should not find error for non-existing name', function() {
        require('should').not.exist(errors.find('FatalDDBError'));
    });
});

describe('errors unique error code generation', function() {

    it('should create unique error codes', function() {
        (fatalError.code).should.not.equal(fatalDBError.code);
        (fatalDBError.code).should.not.equal(fatalDBTransError.code);
    });

    it('should not hammer existing error code', function() {
        var FileNotFoundError = errors.create({
            name: 'FileNotFoundError',
            code: fatalDBTransError.code + 1
        });
        var IOError = errors.create({
            name: 'IOError'
        });

        var fnfError = new FileNotFoundError()
            , ioError = new IOError();

        (fnfError.code).should.not.equal(ioError.code);
        ioError.code.should.be.above(fatalDBTransError.code);
    });
});

describe('default error message handling', function() {

    var FileEncodingError = errors.create({
        name: 'FileEncodingError',
        message: 'File encoding is invalid and cannot be read.'
    });
    var feError = new FileEncodingError();

    it('should use default message when none specified', function() {
        feError.message.should.eql('File encoding is invalid and cannot be read.');
    });

    it('should allow default message overriding', function() {
        var err = new FileEncodingError('dude, the encoding is bad');
        err.message.should.eql('dude, the encoding is bad');
    });
});

describe('scoped creation', function() {
    var MalformedInputError = errors.create({
        name: 'MalformedInputError',
        scope: exports
    });

    it('should exist in exports', function() {
        require('should').exist(exports.MalformedInputError);
        exports.MalformedInputError.should.equal(MalformedInputError);
    });

    it('should find error in exports', function() {
        errors.find('MalformedInputError').should.equal(MalformedInputError);
        var err = new MalformedInputError();
        errors.find(err.code).should.equal(MalformedInputError);
    });
});

describe('errors.stacks()', function() {
    var err = new errors.Http413Error();

    it('should enable stack traces', function() {
        errors.stacks(true);
        err.toString().should.include(err.stack);
    });

    it('should return current value of stacks', function() {
        errors.stacks().should.equal(true);
    });

    it('should disable stack traces', function() {
        errors.stacks(false);
        err.toString().should.not.include(err.stack);
    });

    it('should return current value of stacks', function() {
        errors.stacks().should.equal(false);
    });
});

describe('options style constructor', function() {
    var IdentifiableError = errors.create({name: 'IdentifiableError'}),
        err = new IdentifiableError({message: 'Error with ref ID',
            status: 501, refID: 'a1b2c3'});

    it('should contain refID property', function() {
        err.refID.should.equal('a1b2c3');
    });

    it('should have overridden status', function() {
        err.status.should.equal(501);
    });

    it('toString() should output refID', function() {
        err.toString().should.include('refID: a1b2c3');
    });

    it('toJSON() should include refID', function() {
        err.toJSON().should.include({refID: 'a1b2c3'});
    });

    it('should have overriden message', function() {
        err.toString().should.include(': Error with ref ID');
    });

    it('should not allow overriding of stack', function() {
        (function() {new IdentifiableError({stack: 'fail'});}).should.throw();
    });

    it('should not allow overriding of name', function() {
        (function() {new IdentifiableError({name: 'fail'});}).should.throw();
    });

    it('should not allow overriding of code', function() {
        (function() {new IdentifiableError({code: 601});}).should.throw();
    });
});

describe('status code override', function() {
    var CustomHttpError = errors.create({name: 'CustomHttpError', status: 409}),
        err = new CustomHttpError();

    it('should have status of 409', function() {
        err.status.should.equal(409);
    });

    it('should include 409 status in toJSON()', function() {
        err.toJSON().should.include({status: 409});
    });

    it('should allow overriding in constructor', function() {
        new CustomHttpError({status:411}).status.should.equal(411);
    });
});

describe('errors.isError()', function(){
    var legitError = new errors.HttpError();

    it('should return true for legit errors', function() {
        errors.isError(legitError).should.be.ok;
    });

    it('should return false for other objects', function() {
        errors.isError({}).should.not.be.ok;
    });
});

describe('native error to JSON', function() {
	var NativeError = errors.create({name: 'NativeError'}),
		typeError = new TypeError("Invalid type"),
		useStacks = errors.stacks();


    it('should include basic error attrs', function() {
	   errors.stacks(false);

		var err = errors.errorToJSON(typeError);
		err.should.include({name: 'TypeError', message: 'Invalid type'});
		err.should.not.have.property('stack');

        errors.stacks(useStacks);
    });

    it('should include stack', function() {
        errors.stacks(true);

        var err = errors.errorToJSON(typeError);
        err.should.have.property('stack');
        err.stack.should.equal(typeError.stack);

        errors.stacks(useStacks);
    });

    it('should remap error attrs', function() {
        errors.stacks(true);

        var err = errors.errorToJSON(typeError, {'theStack': ['stack']});
        err.should.have.property('theStack');
        err.theStack.should.equal(typeError.stack);

        errors.stacks(useStacks);
	});

	it('should map nested object properties', function() {
		var typeError = new TypeError("Invalid type");
		typeError['causes'] = {'unknown': 'unknown native cause'};
		var err = errors.errorToJSON(typeError, {'cause': ['causes.unknown']});
		err.cause.should.equal('unknown native cause');
	});
});

describe('user defined attrs', function() {
    var TestError = errors.create({
            name: 'TestError',
            fields: {},
        }),
        test1Error = new TestError(),
        test2Error = new TestError({ fields: { a: 1 } });


    it('should include default attrs', function() {
        test1Error.should.include({name: 'TestError',fields: {}});
    });

    it('should include user defined attrs', function() {
        test2Error.should.include({name: 'TestError', fields: { a: 1 } });
    });

    it('should update user defined attrs on demand', function() {
        test2Error.fields.b = 2;
        test2Error.should.include({name: 'TestError', fields: { a: 1, b: 2 } });
    });

    it('should include user defined attrs in JSON', function() {
        const err = test2Error.toJSON();
        err.should.not.have.property('stack');
        err.should.include({ fields: { a: 1, b: 2 } });
    });

});


test/flow/BaseError.js
/* @flow */

// =============================================================================
// Imports
// =============================================================================
import Errors, { type ErrorsModuleType } from '../../src/errors';

// Flow types
// --------------------------------------------------------
export type BaseErrorOptionsType = {
    _message?: string,
    myProp?: number,
};
/* To allow any prop
type BaseErrorOptionsType = Object;
*/


// =============================================================================
// Constants
// =============================================================================
const BaseErrors: ErrorsModuleType<BaseErrorOptionsType> = Errors;


// =============================================================================
// BaseError
// =============================================================================
export const BaseError = BaseErrors.create({
    name: 'BaseError',
    _message: 'some message',
    myProp: 1,
});

export default BaseErrors;


// =============================================================================
// Tests
// =============================================================================

// Correct
// --------------------------------------------------------
const baseError = new BaseError();

baseError._message
baseError.myProp

baseError.code
baseError.status
baseError.response
baseError.explanation

// Errors
// --------------------------------------------------------
// $FlowThrowsError: _message1 is unknown
baseError._message1

BaseErrors.create({
    name: 'BaseError',
    // $FlowThrowsError: _message must be a string
    _message: 1,
});

BaseErrors.create({
    name: 'BaseError',
    // $FlowThrowsError: myProp must be a number
    myProp: '1',
});


examples/basic/usage.js

var errors = require('../..');

// barebones
errors.create({name: 'RuntimeError'});
console.log(new errors.RuntimeError().toString());

console.log("-------------");

// default message
errors.create({
    name: 'RuntimeError',
    message: 'A runtime error occured during processing'
});
console.log(new errors.RuntimeError().toString());

console.log("-------------");

// default message, explanation and response
errors.create({
    name: 'FileNotFoundError',
    message: 'The requested file could not be found',
    explanation: 'The file /home/boden/foo could not be found',
    response: 'Verify the file exists and retry the operation'
});
console.log(new errors.FileNotFoundError().toString());

console.log("-------------");

// override messages
console.log(new errors.FileNotFoundError(
        'Cannot read file'
        , 'You do not have access to read /root/foo'
        , 'Request a file you have permissions to access').toString());

console.log("-------------");

// define code
errors.create({
    name: 'SecurityError',
    code: 1100
});
console.log(new errors.SecurityError().toString());

console.log("-------------");

// inheritance
errors.create({
    name: 'FatalError',
    message: 'A fatal error occurred',
});
errors.create({
    name: 'FatalSecurityError',
    message: 'A security error occurred, the app must exit',
    parent: errors.FatalError
});
try {
    throw new errors.FatalSecurityError();
} catch (e) {
    if (e instanceof errors.FatalError) {
        // exit
        console.log("Application is shutting down...");
    }
}

console.log("-------------");

// namespace
errors.create({
    name: 'MalformedExpressionError',
    scope: exports
});
console.log(new exports.MalformedExpressionError().toString());

console.log("-------------");

// lookup
if (errors.find(1100) === errors.find('SecurityError')) {
    console.log("Found em");
}

console.log("-------------");

// mappers
errors.create({name: 'InvalidUsernameError'});
errors.create({name: 'InvalidPasswordError'});
errors.mapper(['InvalidUsernameError', 'InvalidPasswordError'], function(err) {
    return new errors.SecurityError('Invalid credentials supplied');
});
console.log(errors.mapError(new errors.InvalidUsernameError()).toString());
console.log(errors.mapError(new errors.InvalidPasswordError()).toString());

console.log("-------------");

// stacks
errors.stacks(true);
console.log(new errors.Http413Error().toString());
console.log(new errors.Http413Error().toJSON());


console.log("---------------");

// options style constructor
errors.create({name:'IdentifiableError'});
console.log(new errors.IdentifiableError({
	message: "Unexpected error",
	refID: "a1b2c3",
	status: 500}).toString());
console.log(new errors.IdentifiableError({
	message: "Unexpected error",
	refID: "a1b2c3",
	status: 500}).toJSON());


console.log("---------------");

// override status in proto
errors.create({name: 'CustomHttpError', status: 488});
console.log(new errors.CustomHttpError().toString());
console.log(new errors.CustomHttpError().toJSON());
